language/compiler/ir_to_bytecode/src/compiler.rs

```diff
@@ -1547,52 +1571,79 @@ impl<S: Scope + Sized> Compiler<S> {
+        match t {
+            Type::Address => Ok(SignatureToken::Address),
+            Type::U64 => Ok(SignatureToken::U64),
+            Type::Bool => Ok(SignatureToken::Bool),
+            Type::ByteArray => Ok(SignatureToken::ByteArray),
+            Type::Reference(is_mutable, inner_type) => {
+                let inner_token = Box::new(self.build_signature_token(inner_type)?);
                 if *is_mutable {
                     Ok(SignatureToken::MutableReference(inner_token))
                 } else {
                     Ok(SignatureToken::Reference(inner_token))
                 }
             }
-        }
-    }
-
-    fn build_normal_signature_token(
-        &mut self,
-        kind: &ast::Kind,
-        tag: &Tag,
-    ) -> Result<SignatureToken> {
-        match (kind, tag) {
-            (ast::Kind::Value, Tag::Address) => Ok(SignatureToken::Address),
-            (ast::Kind::Value, Tag::U64) => Ok(SignatureToken::U64),
-            (ast::Kind::Value, Tag::Bool) => Ok(SignatureToken::Bool),
-            (ast::Kind::Value, Tag::ByteArray) => Ok(SignatureToken::ByteArray),
-            (kind, Tag::Struct(ctype)) => {
-                let module_name = &ctype.module().name();
-                let module_idx = if self.scope.get_name().is_ok() && module_name == ModuleName::SELF
-                {
-                    ModuleHandleIndex::new(0)
-                } else {
-                    self.scope.get_imported_module_handle(module_name)?
-                };
-                let name_idx = self.make_string(&ctype.name().name_ref())?;
-                let kind = match kind {
-                    ast::Kind::Value => Kind::Copyable,
-                    ast::Kind::Resource => Kind::Resource,
-                };
+            Type::Struct(ident) => {
+                // TODO: the following implementation is
+                //         1) a little bit hacky -- APIs leak abstractions
+                //         2) inefficient (linear time look up)
+                //       fix these two problems
+                let module_name = ident.module().name_ref();
+                let struct_name = ident.name().name_ref();
+
+                // in order to know the kind of the struct, we need to perform a look up using the
+                // qualified name of the struct (Module.Name)
+                let (module_idx, kind) =
+                    if self.scope.get_name().is_ok() && module_name == ModuleName::SELF {
+                        // case 1: the struct is defined in the module currently being compiled
+                        //         in this case, we look up the struct handles and find the one
+                        //         with matching module (self) & name
+                        //         we look up the definitions as they may not be ready at the time
+                        // this         function is called
+                        let mut kind = None;
+                        for sh in self.scope.get_struct_handles() {
```

Comment:
As suggested by your comment, this should really be a map, roughly `Map<&str, Map<&str, &struct handle>`, then you won't need this loop